addons/mass_mailing/models/mailing.py

```diff
@@ -1306,35 +1312,86 @@ def _unsubscribe_token(self, res_id, email):
         token = (self.env.cr.dbname, self.id, int(res_id), tools.ustr(email))
         return hmac.new(secret.encode('utf-8'), repr(token).encode('utf-8'), hashlib.sha512).hexdigest()
 
+
+    def _image_to_url(self, b64image: bytes):
+        """Store an image in an attachement and returns an url"""
+        attachment = self.env['ir.attachment'].create({
+            'datas': b64image,
+            'name': "cropped_image_mailing_{}".format(self.id),
+            'type': 'binary',})
+
+        attachment.generate_access_token()
+
+        return '/web/image/%s?access_token=%s' % (
+            attachment.id, attachment.access_token)
+
+    def _get_image_from_url(self, url):
+        head = requests.head(url, allow_redirects=True, timeout=5)
+        if int(head.headers.get('Content-Length', 0)) <= 10 * 1024 * 1024:  # 10MB
```

Comment:
Maybe check as well that the content-type contains `img`? (we probably don't want svg here). 
In `base_import` we have `int(config.get("import_image_maxbytes", DEFAULT_IMAGE_MAXBYTES))` that can be reused here.

Reply:
Something like this then? I copied `_import_image_by_url` with minimal adaptations.
```diff
diff --git a/addons/mass_mailing/models/mailing.py b/addons/mass_mailing/models/mailing.py
index 3bbd9479481..91aae68463e 100644
--- a/addons/mass_mailing/models/mailing.py
+++ b/addons/mass_mailing/models/mailing.py
@@ -13,11 +13,13 @@ from ast import literal_eval
 from dateutil.relativedelta import relativedelta
 from markupsafe import Markup
 from werkzeug.urls import url_join
-from PIL import UnidentifiedImageError
+from PIL import Image, UnidentifiedImageError
 import requests
 import base64
+import io
 
 from odoo import api, fields, models, tools, _
+from odoo.addons.base_import.models.base_import import ImportValidationError
 from odoo.exceptions import UserError, ValidationError
 from odoo.osv import expression
 
@@ -26,7 +28,9 @@ _logger = logging.getLogger(__name__)
 # Syntax of the data URL Scheme: https://tools.ietf.org/html/rfc2397#section-3
 # Used to find inline images
 image_re = re.compile(r"data:(image/[A-Za-z]+);base64,(.*)")
+DEFAULT_IMAGE_TIMEOUT = 3
 DEFAULT_IMAGE_MAXBYTES = 10 * 1024 * 1024  # 10MB
+DEFAULT_IMAGE_CHUNK_SIZE = 32768
 
 mso_re = re.compile(r"\[if mso\]>[\s\S]*<!\[endif\]")
 
@@ -1326,14 +1330,42 @@ class MassMailing(models.Model):
         return '/web/image/%s?access_token=%s' % (
             attachment.id, attachment.access_token)
 
-    def _get_image_from_url(self, url):
-        head = requests.head(url, allow_redirects=True, timeout=5)
+    def _get_image_from_url(self, url, session):
         maxsize = int(tools.config.get("import_image_maxbytes", DEFAULT_IMAGE_MAXBYTES))
-        if (
-            int(head.headers.get('Content-Length', 0)) <= maxsize and
-            'image' in head.headers.get('Content-Type', 'text/html')
-        ):
-            return requests.get(url).content
+        _logger.debug("Trying to import image from URL: %s" % url)
+        try:
+            response = session.get(url, timeout=int(tools.config.get("import_image_timeout", DEFAULT_IMAGE_TIMEOUT)))
+            response.raise_for_status()
+
+            if response.headers.get('Content-Length') and int(response.headers['Content-Length']) > maxsize:
+                raise ImportValidationError(
+                    _("File size exceeds configured maximum (%s bytes)", maxsize)
+                )
+
+            content = bytearray()
+            for chunk in response.iter_content(DEFAULT_IMAGE_CHUNK_SIZE):
+                content += chunk
+                if len(content) > maxsize:
+                    raise ImportValidationError(
+                        _("File size exceeds configured maximum (%s bytes)", maxsize)
+                    )
+
+            image = Image.open(io.BytesIO(content))
+            w, h = image.size
+            if w * h > 42e6:  # Nokia Lumia 1020 photo resolution
+                raise ImportValidationError(
+                    _("Image size excessive, imported images must be smaller than 42 million pixel")
+                )
+
+            return content
+        except Exception as e:
+            _logger.exception(e)
+            raise ImportValidationError(
+                _(
+                    "Could not retrieve URL: %(url)s: %(error)s",
+                    url=url, error=e
+                )
+            )
 
     def _convert_inline_images_to_urls(self, body_html):
         """
@@ -1376,14 +1408,15 @@ class MassMailing(models.Model):
                     absolute_url = url
                     if url.startswith('/'):
                         absolute_url = self.get_base_url() + url
-                    image = self._get_image_from_url(absolute_url)
-                    if image:
-                        width = float(re.search(r'width:\s*([0-9\.]+)\s*px', match).group(1))
-                        height = float(re.search(r'height:\s*([0-9\.]+)\s*px', match).group(1))
-                        image_processor = tools.ImageProcess(image)
-                        image = image_processor.crop_resize(width, height, 0, 0)
-                        comment.text = comment.text.replace(url, self._image_to_url(base64.b64encode(image.source)))
-                        modified = True
+                    with requests.Session() as session:
+                        image = self._get_image_from_url(absolute_url, session)
+                        if image:
+                            target_width = float(re.search(r'width:\s*([0-9\.]+)\s*px', match).group(1))
+                            target_height = float(re.search(r'height:\s*([0-9\.]+)\s*px', match).group(1))
+                            image_processor = tools.ImageProcess(image)
+                            image = image_processor.crop_resize(target_width, target_height, 0, 0)
+                            comment.text = comment.text.replace(url, self._image_to_url(base64.b64encode(image.source)))
+                            modified = True
                 except (ValueError, UnidentifiedImageError, AttributeError):
                     # - ValueError is raised when the url is not a valid url
                     # - UnidentifiedImageError is raised when the url is not
diff --git a/addons/mass_mailing/tests/test_mailing_internals.py b/addons/mass_mailing/tests/test_mailing_internals.py
index a703e6889e2..e1966da7070 100644
--- a/addons/mass_mailing/tests/test_mailing_internals.py
+++ b/addons/mass_mailing/tests/test_mailing_internals.py
@@ -71,7 +71,7 @@ class TestMassMailValues(MassMailCommon):
         # well. We therefore need to crop the image in the VML version to mimick
         # the style of other email clients.
         attachment = {}
-        def patched_get_image(self, url):
+        def patched_get_image(self, url, session):
             return base64.b64decode('iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNk+A8AAQUBAScY42YAAAAASUVORK5CYII=')
         original_image_to_url = self.env['mailing.mailing']._image_to_url
         def patched_image_to_url(self, b64image: bytes):
```